src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs

```diff
@@ -950,14 +962,111 @@ public bool HasSameRules(TimeZoneInfo other)
                 adjustmentRules,
-                disableDaylightSavingTime);
+                disableDaylightSavingTime,
+                hasIanaId);
+        }
+
+        /// <summary>
+        /// Tries to convert IANA time zone Id to Windows Id.
+        /// </summary>
+        /// <param name="ianaId">The IANA time zone Id.</param>
+        /// <param name="windowsId">String object hold the Windows Id which resulted from the IANA Id conversion.</param>
+        /// <returns>True if the Id conversion succeed, false otherwise .</returns>
+        public static unsafe bool TryConvertIanaIdToWindowsId(string ianaId, out string? windowsId)
+        {
+            windowsId = null;
+
+            // This functionality is not enabled in the browser for the sake of size reduction.
+#if !TARGET_BROWSER
+            if (GlobalizationMode.Invariant || GlobalizationMode.UseNls || ianaId is null)
+            {
+                return false;
+            }
+
+            foreach (char c in ianaId)
+            {
+                // ICU uses some characters as a separator and trim the id at that character.
+                // while we should fail if the Id contained one of these characters.
+                if (c == '\\' || c == '\n' || c == '\r')
+                {
+                    return false;
+                }
+            }
+
+            char* buffer = stackalloc char[100];
+            int length = Interop.Globalization.IanaIdToWindowsId(ianaId, buffer, 100);
+            if (length > 0)
+            {
+                windowsId = new string(buffer, 0, length);
+                return true;
+            }
+#endif // !TARGET_BROWSER
+
+            return false;
+        }
+
+        /// <summary>
+        /// Tries to convert Windows time zone Id to IANA Id.
+        /// </summary>
+        /// <param name="windowsId">The Windows time zone Id.</param>
+        /// <param name="ianaId">String object hold the IANA Id which resulted from the Windows Id conversion.</param>
+        /// <returns>True if the Id conversion succeed, false otherwise .</returns>
+        public static bool TryConvertWindowsIdToIanaId(string windowsId, out string? ianaId) => TryConvertWindowsIdToIanaId(windowsId, null, out ianaId);
+
+        /// <summary>
+        /// Tries to convert Windows time zone Id to IANA Id.
+        /// </summary>
+        /// <param name="windowsId">The Windows time zone Id.</param>
+        /// <param name="region">The ISO 3166 for the country/region.</param>
+        /// <param name="ianaId">String object hold the IANA Id which resulted from the Windows Id conversion.</param>
+        /// <returns>True if the Id conversion succeed, false otherwise .</returns>
+        public static unsafe bool TryConvertWindowsIdToIanaId(string windowsId, string? region, out string? ianaId)
+        {
+            ianaId = null;
+
+            // This functionality is not enabled in the browser for the sake of size reduction.
+#if !TARGET_BROWSER
```

Comment:
Can we refactor these `#if`s out into separate partial files? We've been trying hard not to use `#if` in these files.

Reply:
That will introduce a new file for just this small case. currently we have many TimeZoneInfo.?????.cs files which became very confusing. let me know if you still think it is worth splitting that to a new file.

Reply:
See https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/project-guidelines.md#conventions-for-forked-code.

> In general we prefer different code files over #ifdef's because it forces us to better factor the code which leads to easier maintenance over time.

Reply:
Yes, I know that. we already using `#if TARGET_BROWSER` in other places too for the case we think it is not worth splitting to a file. anyway, I'll move it to its own file. do you have suggestion for the file name? `TimeZoneInfo.NonBrowser.cs`?

Reply:
I think we could reuse the `FullGlobalizationData` and `MinimalGlobalizationData` files. The `FullGlobalizationData` may need to get split into `Unix` and `Windows` though....

Reply:
> The FullGlobalizationData may need to get split into Unix and Windows though....

That will require me duplicating the code in both files if we do that. We have to have a file that compile for Linux and Windows and not for the browser.

Reply:
I meant:

* FullGlobalizationData.cs
* FullGlobalizationData.Windows.cs
* FullGlobalizationData.Unix.cs

Note this is just an idea.... You can employ whatever naming you think fits the situation best.

Reply:
I have moved the code to its own file. 

TimeZoneInfo.FullGlobalizationData.cs is compiled for all platforms but the browser.
TimeZoneInfo.FullGlobalizationData.Unix.cs is compiled for Unix platforms but the browser.
TimeZoneInfo.MinimalGlobalizationData.cs is compiled for all the browser only.

When code reviewing this, `git` of course get confused. basically, I renamed the original TimeZoneInfo.FullGlobalizationData.cs to TimeZoneInfo.FullGlobalizationData.Unix.cs without any change. And now TimeZoneInfo.FullGlobalizationData.cs just have only the new code I added.

Reply:
@eerhardt thanks for all the feedback. it is helpful. Please let me know if you have any more comments.

Reply:
This is done now.